Fixed #32983 -- Made related_name on symmetric field raise new system check warning.

partofthething opened the pull request:
Since ManyToManyFields defined with `symmetrical=True` do not add a
related field to the target model, including a `related_name` argument
will never do what the coder likely expects. This makes including
a related_name with a symmetrical model raise a system check warning.

ticket-32983

carltongibson reviewed (commented):
Hi @partofthething — thanks for this. 

As per the comment on the ticket, can we investigate a system check here, similar to the existing related field checks: https://docs.djangoproject.com/en/3.2/ref/checks/#related-fields 👍

partofthething commented:
Great idea thanks. I've updated the PR accordingly.

carltongibson reviewed (commented):
Hi @partofthething — good work thank you. 

Looks good. I left a few inline comments. 

It looks like the changes in `tests/field_deconstruction/tests.py` and `tests/model_meta/models.py` can be made as an initial commit: 

`Refs #32983 -- Removed unneeded related_name from test model definitions.`

Then the fix can be applied afterwards, `Fixed #32983 -- Added system check for redundant related_name on symmetrical M2M fields.`

Are you happy to rebase? (If not I can do it) 

Thanks, and Welcome aboard! ⛵️

felixxm commented on docs/ref/checks.txt:
```diff
@@ -313,6 +313,8 @@ Related fields
   with a ``through`` model.
 * **fields.W344**: The field's intermediary table ``<table name>`` clashes with
   the table name of ``<model>``/``<model>.<field name>``.
+* **fields.W345**: ``related_name`` has no effect on ``ManyToManyField``
+  with a symmetrical relationship.
```
Also, wrap at 79 chars.

felixxm commented on docs/ref/models/fields.txt:
```diff
@@ -1828,7 +1828,8 @@ that control how the relationship functions.
     ``person_set`` attribute to the ``Person`` class. Instead, the
     :class:`ManyToManyField` is assumed to be symmetrical -- that is, if I am
-    your friend, then you are my friend.
+    your friend, then you are my friend. Defining a ``related_name`` in this
+    scenario will raise a warning.
```
Agreed, please revert it.

felixxm commented on tests/invalid_models_tests/test_relative_fields.py:
```diff
@@ -128,6 +128,24 @@ class ThroughModel(models.Model):
+    def test_many_to_many_with_useless_related_name(self):
+
+        class ModelM2M(models.Model):
+            m2m = models.ManyToManyField(
+                "self",
+                related_name="children"
+            )
```
Also, this can be single-lined.

partofthething commented on django/db/models/fields/related.py:
```diff
@@ -1258,6 +1258,16 @@ def _check_ignored_options(self, **kwargs):
+        if self.remote_field.symmetrical and self._related_name:
+            warnings.append(
+                checks.Warning(
+                    f'related_name has no effect on ManyToManyField '
+                    'with a symmetrical relationship.',
```
Good points. I put `(e.g. to "self")` because I'm under the impression that you can have a symmetrical relationship to a non-self model as well (probably pretty rare).

partofthething commented:
Thanks for the good review. I believe I have addressed all the comments. I did the rebase as suggested too. Happy to change to the suggested message rather than my slight pertubation on it if you prefer (the `(e.g. to "self")` thing).

carltongibson commented on django/db/models/fields/related.py:
```diff
@@ -1258,6 +1258,16 @@ def _check_ignored_options(self, **kwargs):
+            warnings.append(
+                checks.Warning(
+                    'related_name has no effect on ManyToManyField '
+                    'with a symmetrical relationship (e.g. to "self").',
```
Hey @partofthething — I agree about `e.g. to "self" but can I just suggest: 

```suggestion
                    'with a symmetrical relationship, e.g. to "self".',
```

— rather than the parentheses? 

(Happy to do this if you've run out of steam 😀)

partofthething replied:
sounds good. on it.

carltongibson reviewed (approved):
OK, super. Thanks @partofthething! 🏅
